### PR TITLE
nydusify: introduce chunkdict generate subcommand

### PR DESCRIPTION
**Pull request by newthifans**
###Notice
Please take a look and give me some code suggestions in third commit, thanks!

### Brief introduction
Add the function of storing multiple nydus images into the database for generating chunkdict
Implement the command “nydusify chunkdict generate --sources”
by invoking the command “nydus-image chunkdict save --bootstrap” .

 ### **Basic usage**
nydusify chunkdict generate --sources localhost:5000/ubuntu:nydus_latest,localhost:5000/ubuntu:nydus_latest_v1

### Details
1. Write command “nydusify chunkdict generate” entries
2. Create a “generater” folder to implement the function
3. Store the image in the registry with slices and Store them in the database in turn
4. Parse nydus image and create a directory to hold corresponding bootstrap
5. Pull nydus image bootstrap and store  corresponding directory
6. Invoke  the command “nydus-image chunkdict save --bootstrap”

This PR is related to #1305 

###Notice
This PR only modifies 6 files which is located at “contrib/nydusify”

### CODE REVIEW

**Comment by imeoer** [comment]
@newthifans Please rename the PR title to make it meaningful, and fix the code conflicts, thanks!

**Comment by newthifans** [comment]
> @newthifans Please rename the PR title to make it meaningful, and fix the code conflicts, thanks!

Thanks，I have renamed the PR title. About code conflicts, this PR's first two commits came from PR #1305. The conflicts and bugs of the first two commits will be fixed in #1305. The third commit is my code. Please take a look and give me some code suggestions, thanks!

File: Cargo.toml [hunk 1]
```diff
@@ -57,6 +57,7 @@ openssl = { version = "0.10.48", features = ["vendored"] }
 # pin openssl-src to bring in fix for https://rustsec.org/advisories/RUSTSEC-2022-0032
 #openssl-src = { version = "111.22" }
 
+
```
Comment (imeoer) [inline Cargo.toml:60]: Why is there a such commit?

File: contrib/nydusify/cmd/nydusify.go [hunk 2]
```diff
@@ -629,6 +630,75 @@ func main() {
...
+						&cli.BoolFlag{
+							Name:     "source-insecure",
+							Required: false,
+							Usage:    "Skip verifying server certs for HTTPS source registry",
+							EnvVars:  []string{"SOURCE_INSECURE"},
+						},
+						&cli.StringFlag{
+							Name:    "work-dir",
+							Value:   "./output",
+							Usage:   "Working directory for image verification",
```
Comment (imeoer) [inline contrib/nydusify/cmd/nydusify.go:659]: Why `image verification`?

File: contrib/nydusify/cmd/nydusify.go [hunk 3]
```diff
@@ -629,6 +630,75 @@ func main() {
...
+							EnvVars: []string{"WORK_DIR"},
+						},
+						&cli.StringFlag{
+							Name:    "nydus-image",
+							Value:   "nydus-image",
+							Usage:   "Path to the nydus-image binary, default to search in PATH",
+							EnvVars: []string{"NYDUS_IMAGE"},
+						},
+						&cli.StringFlag{
+							Name:    "nydusd",
```
Comment (imeoer) [inline contrib/nydusify/cmd/nydusify.go:669]: Do we depend on this component?

File: contrib/nydusify/go.mod [hunk 4]
```diff
@@ -29,6 +28,8 @@ require (
 	lukechampine.com/blake3 v1.1.5
 )
 
+require github.com/goharbor/acceleration-service v0.2.4
```
Comment (imeoer) [inline contrib/nydusify/go.mod:31]: Can we move it to the above?

File: contrib/nydusify/pkg/generater/bootstrap.go [hunk 5]
```diff
@@ -0,0 +1,35 @@
+// Copyright 2020 Ant Group. All rights reserved.
```
Comment (imeoer) [inline contrib/nydusify/pkg/generater/bootstrap.go:1]: Copyright 2023 Nydus Developers. All rights reserved.

File: contrib/nydusify/pkg/generater/generater.go [hunk 6]
```diff
@@ -0,0 +1,115 @@
+package generater
```
Comment (imeoer) [inline contrib/nydusify/pkg/generater/generater.go:1]: Can we rename the package to `chunkdict`?

File: contrib/nydusify/pkg/generater/generater.go [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+	NydusImagePath string
+	NydusdPath     string
+	ExpectedArch   string
+}
+
+// Chunkdict generate deposit multiple Nydus images into database
+// by Nydus-image to save image chunk and blob into database.
+type Generater struct {
+	Opt
+	sourceSlice  []string
```
Comment (imeoer) [inline contrib/nydusify/pkg/generater/generater.go:33]: `sources`

File: contrib/nydusify/pkg/generater/generater.go [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+
+// "Chunkdict Generate" deposit multiple Nydus images to the database in turn
+func (generater *Generater) Generate(ctx context.Context) error {
+	for index := 0; index < len(generater.sourceParser); index++ {
+		if err := generater.generate(ctx, index); err != nil {
+			if utils.RetryWithHTTP(err) {
+				generater.sourceParser[index].Remote.MaybeWithHTTP(err)
+			}
+		}
+		generater.generate(ctx, index)
```
Comment (imeoer) [inline contrib/nydusify/pkg/generater/generater.go:71]: The error should be handled.

File: contrib/nydusify/pkg/generater/generater.go [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+	return nil
+}
+
+// A Nydus image is stored in the database
+func (generater *Generater) generate(ctx context.Context, index int) error {
+	sourceParsed, err := generater.sourceParser[index].Parse(ctx)
+	if err != nil {
+		return errors.Wrap(err, "parse Nydus image")
+	}
+	//create a directory to store the image bootstrap
```
Comment (imeoer) [inline contrib/nydusify/pkg/generater/generater.go:82]: Add a space after `//`.

File: contrib/nydusify/pkg/generater/generater.go [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+	nydusImageName := generater.sourceSlice[index][strings.LastIndex(generater.sourceSlice[index], "/")+1:]
+	folderPath := filepath.Join(generater.WorkDir, nydusImageName)
+	_, err = os.ReadDir(folderPath)
```
Comment (imeoer) [inline contrib/nydusify/pkg/generater/generater.go:85]: Should use `os.Stat`?

File: contrib/nydusify/pkg/generater/generater.go [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+	if err == nil {
+		if err := os.RemoveAll(folderPath); err != nil {
+			return errors.Wrap(err, "error delete directory")
```
Comment (imeoer) [inline contrib/nydusify/pkg/generater/generater.go:88]: return errors.Wrap(err, "remove work directory")

File: contrib/nydusify/pkg/generater/generater.go [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+		}
+		if err := os.Mkdir(folderPath, fs.ModePerm); err != nil {
+			return errors.Wrap(err, "error creat directory")
```
Comment (imeoer) [inline contrib/nydusify/pkg/generater/generater.go:91]: return errors.Wrap(err, "create work directory")

File: contrib/nydusify/pkg/generater/tool/builder.go [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+
+func NewBuilder(binaryPath string) *Builder {
+	return &Builder{
+		binaryPath: binaryPath,
+		stdout:     os.Stdout,
+		stderr:     os.Stderr,
+	}
+}
+
+// Save calls `nydus-image save` to parse Nydus bootstrap
```
Comment (imeoer) [inline contrib/nydusify/pkg/generater/tool/builder.go:31]: `nydus-image chunkdict generate`?

File: contrib/nydusify/pkg/generater/output.go [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+) error {
+	if sourceParsed.Index != nil {
+		if err := prettyDump(
+			sourceParsed.Index,
+			filepath.Join(outputPath, "nydus_index.json"),
+		); err != nil {
+			return errors.Wrap(err, "output nydus index file")
+		}
+	}
+	if sourceParsed.NydusImage != nil {
```
Comment (imeoer) [inline contrib/nydusify/pkg/generater/output.go:36]: Should give a error massage if the source is not a nydus image.

**Comment by imeoer** [comment]
The PR title should be `nydusify: introduce chunkdict generate subcommand`.

File: contrib/nydusify/cmd/nydusify.go [hunk 15]
```diff
@@ -629,6 +630,75 @@ func main() {
 				return checker.Check(context.Background())
 			},
 		},
+		{
+			Name:  "chunkdict",
+			Usage: "Deduplicate RAFS filesystem metadata",
+			Subcommands: []*cli.Command{
+				{
+					Name:  "generate",
+					Usage: "Multi-image save chunk info to a database",
```
Comment (hangvane) [inline contrib/nydusify/cmd/nydusify.go:639]: "Save chunk info of multiple images to a database" ?

File: contrib/nydusify/pkg/generater/generater.go [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+
+	"github.com/dragonflyoss/image-service/contrib/nydusify/pkg/parser"
+	"github.com/dragonflyoss/image-service/contrib/nydusify/pkg/provider"
+	"github.com/dragonflyoss/image-service/contrib/nydusify/pkg/utils"
+)
+
+// Opt defines Chunkdict generate options.
+// Note: sources is one and more Nydus image references.
```
Comment (hangvane) [inline contrib/nydusify/pkg/generater/generater.go:19]: one or more?

File: contrib/nydusify/pkg/generater/generater.go [hunk 17]
```diff
@@ -0,0 +1,115 @@
...
+
+// Chunkdict generate deposit multiple Nydus images into database
+// by Nydus-image to save image chunk and blob into database.
+type Generater struct {
+	Opt
+	sourceSlice  []string
+	sourceParser []*parser.Parser
+}
+
+// New creates Generater instance, sourceSlice is multiple Nydus images reference list.
```
Comment (hangvane) [inline contrib/nydusify/pkg/generater/generater.go:37]: Please consider moving the comment for the variable `sourceSlice` to the line above it.

File: contrib/nydusify/pkg/generater/generater.go [hunk 18]
```diff
@@ -0,0 +1,115 @@
...
+	generater := &Generater{
+		Opt:          opt,
+		sourceSlice:  sourceSlice,
+		sourceParser: sourcesParser,
+	}
+
+	return generater, nil
+}
+
+// "Chunkdict Generate" deposit multiple Nydus images to the database in turn
```
Comment (hangvane) [inline contrib/nydusify/pkg/generater/generater.go:63]: Could you please explain `in turn`?

File: src/bin/nydus-image/deduplicate.rs [hunk 19]
```diff
@@ -0,0 +1,239 @@
...
+
+use anyhow::{Context, Result};
+use nydus_api::ConfigV2;
+use nydus_rafs::builder::Tree;
+use nydus_rafs::metadata::RafsSuper;
+use nydus_storage::device::BlobInfo;
+use nydus_storage::meta::format_blob_features;
+use rusqlite::{params, Connection};
+pub struct Deduplicate {
+    sb: RafsSuper,
```
Comment (hangvane) [inline src/bin/nydus-image/deduplicate.rs:18]: Consider using a clearer variable name.

File: contrib/nydusify/pkg/generater/tool/builder.go (hunk 13 above)
Comment (newthifans) [reply contrib/nydusify/pkg/generater/tool/builder.go]: `nydus-image chunkdict save`  achieve function "store nydus image into database", "nydus-image chunkdict generate" achieve algorithm afterwards.

File: Cargo.toml [hunk 20]
```diff
@@ -58,13 +59,17 @@ openssl = { version = "0.10.48", features = ["vendored"] }
 
 nydus-api = { version = "0.2.2", path = "api", features = ["handler"] }
 nydus-rafs = { version = "0.2.2", path = "rafs", features = ["builder"] }
-nydus-service = { version = "0.2.0", path = "service", features = ["block-device"] }
+nydus-service = { version = "0.2.0", path = "service", features = [
```
Comment (jiangliu) [inline Cargo.toml:62]: Unrelated change？

Comment (newthifans) [reply Cargo.toml:62]: This is the change made by the first two commits from **#1305**. I don't understand this part, I work on the third commit,  just calling the subcommands `nydus-image` chunkdict save ” implemented by the first two commits.

File: contrib/nydusify/pkg/chunkdict/generater/bootstrap.go [hunk 21]
```diff
@@ -0,0 +1,36 @@
...
+	BootstrapPath  string
+	NydusImagePath string
+}
+
+func (bootstrap *Bootstrap) Name() string {
+	return "Bootstrap"
+}
+
+func (bootstrap *Bootstrap) Save() error {
+	logrus.Infof("Save chunk and blob table into database")
```
Comment (imeoer) [inline contrib/nydusify/pkg/chunkdict/generater/bootstrap.go:25]: Next, we need to call `nydus-image chunkdict generate` to generate the chunk dict image directly, so let's remove the info message.

File: contrib/nydusify/pkg/chunkdict/generater/generater.go [hunk 22]
```diff
@@ -0,0 +1,117 @@
...
+	"github.com/dragonflyoss/image-service/contrib/nydusify/pkg/parser"
+	"github.com/dragonflyoss/image-service/contrib/nydusify/pkg/provider"
+	"github.com/dragonflyoss/image-service/contrib/nydusify/pkg/utils"
+)
+
+// Opt defines Chunkdict generate options.
+// Note: sources is one or more Nydus image references.
+type Opt struct {
+	WorkDir        string
+	Sources        string
```
Comment (imeoer) [inline contrib/nydusify/pkg/chunkdict/generater/generater.go:22]: Why not `Sources []string`.

File: contrib/nydusify/pkg/chunkdict/generater/generater.go [hunk 23]
```diff
@@ -0,0 +1,117 @@
...
+	SourceInsecure bool
+	NydusImagePath string
+	ExpectedArch   string
+}
+
+// "Chunkdict generate" command store multiple Nydus images into database
+// by invoking "nydus-image chunkdict save" command
+// to save chunk and blob information image into database.
+type Generater struct {
```
Comment (imeoer) [inline contrib/nydusify/pkg/chunkdict/generater/generater.go:31]: Generator.

File: contrib/nydusify/pkg/chunkdict/generater/generater.go [hunk 24]
```diff
@@ -0,0 +1,117 @@
...
+// Note: sources is one or more Nydus image references.
+type Opt struct {
+	WorkDir        string
+	Sources        string
+	SourceInsecure bool
+	NydusImagePath string
+	ExpectedArch   string
+}
+
+// "Chunkdict generate" command store multiple Nydus images into database
```
Comment (imeoer) [inline contrib/nydusify/pkg/chunkdict/generater/generater.go:28]: `chunkdict generate` should be used for generating the chunk dict image directly.

File: contrib/nydusify/pkg/chunkdict/generater/generater.go [hunk 25]
```diff
@@ -0,0 +1,117 @@
...
+// by invoking "nydus-image chunkdict save" command
+// to save chunk and blob information image into database.
+type Generater struct {
+	Opt
+	// Nydus image references slice.
+	sourcesSlice  []string
```
Comment (imeoer) [inline contrib/nydusify/pkg/chunkdict/generater/generater.go:34]: It's better not to attach a type to a variable name.

File: contrib/nydusify/pkg/chunkdict/generater/generater.go [hunk 26]
```diff
@@ -0,0 +1,117 @@
...
+	var sourcesParser []*parser.Parser
+	for _, source := range sourcesSlice {
+		sourcesRemote, err := provider.DefaultRemote(source, opt.SourceInsecure)
+		if err != nil {
+			return nil, errors.Wrap(err, "Init source image parser")
+		}
+		sourceParser, err := parser.New(sourcesRemote, opt.ExpectedArch)
+		sourcesParser = append(sourcesParser, sourceParser)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to create parser")
```
Comment (imeoer) [inline contrib/nydusify/pkg/chunkdict/generater/generater.go:51]: Please use consistent capitalization in the error message.

File: contrib/nydusify/pkg/chunkdict/generater/generater.go [hunk 27]
```diff
@@ -0,0 +1,117 @@
...
+
+// "Chunkdict Generate" deposit multiple Nydus images into the database one by one.
+func (generater *Generater) Generate(ctx context.Context) error {
+	for index := 0; index < len(generater.sourcesParser); index++ {
+		if err := generater.generate(ctx, index); err != nil {
+			if utils.RetryWithHTTP(err) {
+				generater.sourcesParser[index].Remote.MaybeWithHTTP(err)
+			}
+		}
+		generater.generate(ctx, index)
```
Comment (imeoer) [inline contrib/nydusify/pkg/chunkdict/generater/generater.go:72]: We should not miss any error handling.

File: contrib/nydusify/pkg/chunkdict/generater/generater.go [hunk 28]
```diff
@@ -0,0 +1,117 @@
...
+	}
+	return nil
+}
+
+// A Nydus image is stored in the database
```
Comment (imeoer) [inline contrib/nydusify/pkg/chunkdict/generater/generater.go:77]: Can we explain the generate method clearly?

File: contrib/nydusify/pkg/chunkdict/generater/generater.go [hunk 29]
```diff
@@ -0,0 +1,117 @@
...
+func (generater *Generater) generate(ctx context.Context, index int) error {
+	sourceParsed, err := generater.sourcesParser[index].Parse(ctx)
+	if err != nil {
+		return errors.Wrap(err, "parse Nydus image")
+	}
+
+	// Create a directory to store the image bootstrap
+	nydusImageName := generater.sourcesSlice[index][strings.LastIndex(generater.sourcesSlice[index], "/")+1:]
```
Comment (imeoer) [inline contrib/nydusify/pkg/chunkdict/generater/generater.go:85]: How can we ensure no have conflict name with a different registry host?

File: contrib/nydusify/pkg/chunkdict/generater/generater.go [hunk 30]
```diff
@@ -0,0 +1,117 @@
...
+	folderPath := filepath.Join(generater.WorkDir, nydusImageName)
+	_, err = os.Stat(folderPath)
+	if err == nil {
```
Comment (imeoer) [inline contrib/nydusify/pkg/chunkdict/generater/generater.go:88]: Can we use os.MkdirAll here?
`os.MkdirAll`:  `If path is already a directory, MkdirAll does nothing and returns nil.`

Comment (imeoer) [reply contrib/nydusify/pkg/chunkdict/generater/generater.go:88]: And we need to clean up the work directory after the generation finish.

File: contrib/nydusify/pkg/build/builder.go [hunk 31]
```diff
@@ -143,3 +147,24 @@ func (builder *Builder) Run(option BuilderOption) error {
...
+		option.BootstrapPath,
+	}
+	cmd := exec.Command(builder.binaryPath, args...)
+	cmd.Stdout = builder.stdout
+	cmd.Stderr = builder.stderr
+
+	if err := cmd.Run(); err != nil {
+		return err
+	}
+	return nil
```
Comment (Desiki-high) [inline contrib/nydusify/pkg/build/builder.go:169]: Maybe ```return cmd.Run()```

File: contrib/nydusify/pkg/chunkdict/generator/generator.go [hunk 32]
```diff
@@ -0,0 +1,107 @@
...
+	for index := 0; index < len(generator.sourcesParser); index++ {
+		if err := generator.save(ctx, index); err != nil {
+			if utils.RetryWithHTTP(err) {
+				generator.sourcesParser[index].Remote.MaybeWithHTTP(err)
+			}
+			if err := generator.save(ctx, index); err != nil {
+				return err
+			}
+		}
+	}
```
Comment (Desiki-high) [inline contrib/nydusify/pkg/chunkdict/generator/generator.go:71]: Maybe use ```for index, sourceParser := range generator.sourcesParser``` is better than ```len(generator.sourcesParser)``` ?

File: contrib/nydusify/pkg/chunkdict/generator/generator.go [hunk 33]
```diff
@@ -0,0 +1,107 @@
...
+	return nil
+}
+
+// "Save" function stores information of chunk and blob of a Nydus Image in the database
+func (generator *Generator) save(ctx context.Context, index int) error {
```
Comment (Desiki-high) [inline contrib/nydusify/pkg/chunkdict/generator/generator.go:76]: Does ```index int``` better than ```sourceParser *parser.Parser```? why we must need the index?

Comment (newthifans) [reply contrib/nydusify/pkg/chunkdict/generator/generator.go:76]: `Index int` indicates two  the ordinal numbers of the two arrays `Source` and  `sourceParser`. Using `index` is easier.
